big_tests/tests/carboncopy_helper.erl

```diff
@@ -1,8 +1,80 @@
 -module(carboncopy_helper).
--export([wait_for_carbon_with_body/3]).
+-export([
+    wait_for_carbon_chat_with_body/3,
+    wait_for_carbon_message/2,
+    normal_message_with_body/2,
+    normal_message_with_receipt/2,
+    normal_message_with_csn/2,
+    normal_message_with_chat_marker/2
+]).
 
-wait_for_carbon_with_body(Client, Body, #{from := From, to := To}) ->
+-include_lib("exml/include/exml.hrl").
+-include_lib("escalus/include/escalus_xmlns.hrl").
+
+wait_for_carbon_chat_with_body(Client, Body, #{from := From, to := To}) when is_binary(From) ->
+    escalus:assert(
+        is_forwarded_received_message,
+        [From, escalus_client:full_jid(To), Body],
+        escalus_client:wait_for_stanza(Client)
+    );
+
+wait_for_carbon_chat_with_body(Client, Body, #{from := From, to := To}) ->
     escalus:assert(
-      is_forwarded_received_message,
-      [escalus_client:full_jid(From), escalus_client:full_jid(To), Body],
-      escalus_client:wait_for_stanza(Client)).
+        is_forwarded_received_message,
+        [escalus_client:full_jid(From), escalus_client:full_jid(To), Body],
+        escalus_client:wait_for_stanza(Client)
+    ).
+
+wait_for_carbon_message(Client, #{from := From, to := To}) ->
+    escalus:assert(
+        fun is_forwarded_received_message/3,
+        [escalus_client:full_jid(From), escalus_client:full_jid(To)],
+        escalus_client:wait_for_stanza(Client)
+    ).
+
+is_forwarded_received_message(From, To, Stanza) ->
+    Carbon = exml_query:subelement(Stanza, <<"received">>),
+    escalus_pred:has_ns(?NS_CARBONS_2, Carbon) andalso
+        is_forwarded_message(From, To, exml_query:subelement(Carbon, <<"forwarded">>)).
+
+is_forwarded_message(From, To, Stanza) ->
+    escalus_pred:has_ns(?NS_FORWARD_0, Stanza) andalso
+        is_message_from_to(From, To, exml_query:subelement(Stanza, <<"message">>)).
+
+is_message_from_to(From, To, #xmlel{attrs = Attrs} = Stanza) ->
+    escalus_pred:is_message(Stanza) andalso
+        escalus_compat:bin(From) == proplists:get_value(<<"from">>, Attrs) andalso
+        escalus_compat:bin(To) == proplists:get_value(<<"to">>, Attrs).
+
+normal_message_with_body(User, Body) ->
+    escalus_stanza:message(Body, #{type => <<"normal">>, to => User}).
+
+normal_message_with_receipt(User, _Body) ->
```

Comment:
These functions acept Body but ignore it - a bit strange and unexpected. Maybe solve the arity in a different way?

Reply:
I decided to go with map as an argument and IMO it looks better now. Thanks 🙂 Please have a look if you think that as well.